Add iterator abort

nemild opened the pull request.

ethers commented:
An "iterator abort" is an instance of "DoS with (Unexpected) Throw".  I'll add another example and let's see how we might restructure.

nemild commented:
@ethers That's right. Here are some things I noticed:

- I saw the original title as an *attack*, not a *Recommendation for Smart Contract Security*, so that's why I moved to that section; eventually I was thinking the attacks could each point to a solution in the recommendations section (like the pull vs push payments)

- I tried to cut the amount of text down, though we could always add a sentence describing a common example where this occurs

- Then I added the simplest example I could think about

I'm open to any name that the community feels is the correct one.
